cabby/geo/util.py

```diff
@@ -618,4 +619,63 @@ def point_str_to_shapely_point(point_str: Text) -> Point:
   point_str=point_str.split(')')[0]
   coords = point_str.split(" ")
   x, y = float(coords[0]), float(coords[1])
-  return Point(x,y)
\ No newline at end of file
+  return Point(x,y)
+
+def closest_point_point2polygon(poly_goal: Polygon, point_ref: Point) -> Point:
+  ''' Calculate the closest point on a polygon to a given point.
+  Arguments:
+    poly_goal: The polygon which the closest point should be on.
+    point_ref: The given point outside the polygon.
+  Returns:
+    A closest point on the polygon and the distance.
+  '''
+  pol_ext = LinearRing(poly_goal.exterior.coords)
+  dist = pol_ext.project(point_ref)
+  line = pol_ext.interpolate(dist)
+  return Point(list(line.coords)[0]), dist
+
+def closest_point_linestring2polygon(
+  poly_goal: Polygon, line_ref: LineString) -> Point:
+  ''' Calculate the closest point between a line and a polygon.
+  Arguments:
+    poly_goal: The polygon which the closest point should be on.
+    line_ref: The given line the point on poly1 should be close to.
+  Returns:
+    A closest point on the polygon and distance.
+  '''
+
+  min_dist = np.Inf
+  for coord_ref in line_ref.coords:
+    point_ref = Point(coord_ref)
+    point_found, dist = closest_point_point2polygon(poly_goal, point_ref)
+    if dist < min_dist:
+      min_dist = dist
+      closest_point = point_found
```

Comment:
There is likely a more efficient way to compute the closest point than exhaustively computing all pairwise distances (though there are probably many corner cases). E.g. there might be some existing shapely functions you can call to do this more efficiently as discussed here: https://stackoverflow.com/questions/33311616/find-coordinate-of-the-closest-point-on-polygon-in-shapely

Fine to keep as is, but might be worth revisiting in future if this becomes a bottleneck.